### PR TITLE
enhance gvsbuild

### PR DESCRIPTION
**Pull request by congzhangzh**
1. add vs 2017 support
2. more accurate path variable fetch and merge
3. python 2/3 byte,str problem friend
4. subprocess accurate run and non zero catch for easy debug
5. add CmakeProject a cmake_ext_cmdline parameter to make CmakeProject more flexible
6. add a new GitCmakeProject project type which is needed for pygobject
7. adjust some project to use GitCmakeProject
notes: pyojbect can not build successful

### CODE REVIEW

File: gvsbuild/utils/builder.py [hunk 1]
```diff
@@ -105,10 +112,25 @@ def __check_vs(self, opts):
...
+        try:
+            # Use output redirect will help to ignore the header part vcvars family print
+            # Which make later parser more reliable
+            cmd_details='cmd.exe /c ""%s" && set>%s"' % (vcvars_bat, temp_env_out_path)
+            retcode = subprocess.call(cmd_details, shell=True)
+            if retcode==0:
+                with open(temp_env_out_path) as temp_env_out_file:
+                    output=temp_env_out_file.read()
+            else:
+                raise Exception("Some error happened when try to run vcvars family, cmd: %s, cwd: %s" % (cmd_details, os.getcwd()))
```
Comment (TingPing) [inline gvsbuild/utils/builder.py]: Use `check_call` rather than making your own exception

File: gvsbuild/utils/builder.py [hunk 2]
```diff
@@ -362,7 +384,11 @@ def __execute(self, args, working_dir=None, add_path=None, env=None):
             else:
                 env = dict(os.environ)
             self.__add_path(env, add_path)
-        subprocess.check_call(args, cwd=working_dir, env=env, shell=True)
+        try:
+            subprocess.check_call(args, cwd=working_dir, env=env, shell=True)
+        except Exception as exc:
```
Comment (TingPing) [inline gvsbuild/utils/builder.py]: Use a less vague exception like `CalledProcessError`

**Comment by TingPing** [comment]
Should probably be split into separate commits to be more clear what is being changed.

File: gvsbuild/utils/builder.py (hunk 1 above)
Comment (congzhangzh) [inline gvsbuild/utils/builder.py]: the self provide exception can provide more details for easy debug, the default CalledProcessError like 

 subprocess.CalledProcessError: Command 'exit 1' returned non-zero exit status 1

user does not know where the command be called, the cwd info is very important for debug.

Comment (congzhangzh) [inline gvsbuild/utils/builder.py]: maybe try and catch as CalledProcessError, print_debug some information and raise?

Comment (TingPing) [inline gvsbuild/utils/builder.py]: Yea that seems reasonable.

**Comment by congzhangzh** [comment]
this also fix #122

**Comment by nacho** [comment]
Thanks for working on this. As @tingping pointed out, can u split this into different patches?

**Comment by congzhangzh** [comment]
I suggest not to mind this, we still have a lot of things to do, this is just an enhancement process, Just merge it please:)

**Comment by nacho** [comment]
No please, split the patches

**Comment by nacho** [comment]
Yesterday I've split and pushed a couple of patches

**Comment by TingPing** [comment]
@nacho On a related note, is there any reason to support Python 2 at all?

**Comment by nacho** [comment]
Not really, in fact I think we specify on the readme to use always python3

**Comment by congzhangzh** [comment]
If we mark web as a modern tech and fast evolving tech, GTK is really a old
and a little slow envolving.

2017-06-01 23:36 GMT+08:00 Ignacio Casal Quinteiro <notifications@github.com

**Comment by congzhangzh** [comment]
Sorry, press send button accidentally.

Many old system is base on python 2 family especially for infrastructure
part.

Upgrade whole old system base on python 2 to 3 is a pain,  even google's
infrastructure need python 2 for chromium os.

I know many package owner has already drop python 2 support. but make build
script work with python 2 is not so bad as I think:)


2017-06-02 2:38 GMT+08:00 Cong Monkey <congzhangzh@gmail.com>:

**Comment by congzhangzh** [comment]
And very happy I can do something for gvsbuild:)

What I expect most is python-gi support, As I use pyside in my project
which is too old and PYQT have some license probem, I want give
python-gobject a try in the future:)

2017-06-02 2:52 GMT+08:00 Cong Monkey <congzhangzh@gmail.com>:

**Comment by guruDanny67** [comment]
We are not on Linux when python is used by the system and upgrading the whole system from 2.7 to 3.3 or 3.4 is really a pain and a complex matter (but it's moving).

On window (and you can do it also on linux, in your local space) you can have all the different version of python, 2.7, 3.3, 3.4, 3.5, 3.6, in 32 and 64 bit.

The script doesn't need any package outside of the base installer: you can simply estract the python 3.5 in some directory, say c:\mystuff\test\see_if_it_works\python_3.5.2-32bit, and call the script with that python (c:\mystuff\test\see_if_it_works\python_3.5.2-32bit\python.exe build.py build ...) and it work, even if you are doing all your development with a 2.7 version.

Actually we target 3.4 minimun version (the 3.3 verion and before gots a lot of problems and are better left alone ...) but I normally use the 3.5 version or the 3.6.

Considering that the 3.5 version is the first version builded with vs2015 if we want to build the gobject introspection for python the target version must be that or more: if you want to use our py-gobject with, say, python 3.3 you need to rebuild python with the same compiler used for gobject and it will be a really mess.

For gobject, if you want to use python 2.7-3.4 and don't want a brand new gtk version there is a package on sourceforge.net with a lot of modules called pygobject: you can have a a try of the gtk3 now.

To summarize: being the script so self contained I thing there is no pressure to support python 2.7 (that can be worse than now if we, for example, have/want to use six to mantain compatibility with 2.7/3.x python version).
